virtualcluster/pkg/syncer/resources/pod/mutatorplugin/podkubeapiaccessmutator.go

```diff
@@ -0,0 +1,269 @@
+	uplugin "sigs.k8s.io/cluster-api-provider-nested/virtualcluster/pkg/util/plugin"
+)
+
+const (
+	// DefaultServiceAccountName is the name of the default service account to set on pods which do not specify a service account
+	DefaultServiceAccountName = "default" // #nosec G101
+
+	// ServiceAccountVolumeName is the prefix name that will be added to volumes that mount ServiceAccount secrets
+	ServiceAccountVolumeName = "kube-api-access" // #nosec G101
+
+	// DefaultAPITokenMountPath is the path that ServiceAccountToken secrets are automounted to.
+	// The token file would then be accessible at /var/run/secrets/kubernetes.io/serviceaccount
+	DefaultAPITokenMountPath = "/var/run/secrets/kubernetes.io/serviceaccount" // #nosec G101
+
+	defaultAttemptTimes = 10
+)
+
+func init() {
+	MutatorRegister.Register(&uplugin.Registration{
+		ID: "00_PodKubeAPIAccessMutator",
+		InitFn: func(ctx *uplugin.InitContext) (interface{}, error) {
+			return NewPodKubeAPIAccessMutatorPlugin(ctx)
+		},
+	})
+}
+
+type PodKubeAPIAccessMutatorPlugin struct {
+	client       kubernetes.Interface
+	generateName func(string) string
+}
+
+func NewPodKubeAPIAccessMutatorPlugin(ctx *uplugin.InitContext) (*PodKubeAPIAccessMutatorPlugin, error) {
+	plugin := &PodKubeAPIAccessMutatorPlugin{
+		client:       ctx.Client,
+		generateName: names.SimpleNameGenerator.GenerateName,
+	}
+	return plugin, nil
+}
+
+func (pl *PodKubeAPIAccessMutatorPlugin) Mutator() conversion.PodMutator {
+	return func(p *conversion.PodMutateCtx) error {
+		if !featuregate.DefaultFeatureGate.Enabled(featuregate.KubeAPIAccessSupport) {
+			return nil
+		}
+
+		// Don't modify the spec of mirror pods.
+		// That makes the kubelet very angry and confused, and it immediately deletes the pod (because the spec doesn't match)
+		// That said, don't allow mirror pods to reference ServiceAccounts or SecretVolumeSources either
+		if _, isMirrorPod := p.PPod.Annotations[corev1.MirrorPodAnnotationKey]; isMirrorPod {
+			return nil
+		}
+
+		// Set the default service account if needed
+		if len(p.PPod.Spec.ServiceAccountName) == 0 {
+			p.PPod.Spec.ServiceAccountName = DefaultServiceAccountName
+		}
+
+		targetNamespace := conversion.ToSuperClusterNamespace(p.ClusterName, p.PPod.Namespace)
+		serviceAccount, err := pl.getServiceAccount(targetNamespace, p.PPod.Spec.ServiceAccountName)
+		if err != nil {
+			return fmt.Errorf("error looking up serviceAccount %s/%s: %v", targetNamespace, p.PPod.Spec.ServiceAccountName, err)
+		}
+
+		secret, err := pl.getSecret(p.ClusterName, targetNamespace, serviceAccount)
+		if err != nil || secret == nil {
+			klog.Errorf("not found serviceAccount %s/%s token secret, err: %v", targetNamespace, p.PPod.Spec.ServiceAccountName, err)
+			return fmt.Errorf("error looking up secret of serviceAccount %s/%s: %v", targetNamespace, p.PPod.Spec.ServiceAccountName, err)
+		}
+
+		if shouldAutomount(serviceAccount, p.PPod) {
```

Comment:
Can we use `automountServiceAccountToken` on the [pod spec](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.27/#pod-v1-core)?

Reply:
We explicitly override this from vc-syncer so the value in the pPod should always be false, seems like if we have the vPod object we could get that value.